File: scala/SuffixArray.scala
package com.github.antma.cpalgo

import annotation.tailrec

sealed class SAFindResult
case class SASuffix(val o: Int) extends SAFindResult
case class SAInterval(val d: Int, f: Int) extends SAFindResult

class SuffixArray (_s: String) {
  val s = _s :+ 0.toChar
  val n = _s.length
  private val sigma = (_s.fold (0.toChar) ( (x, y) => x.max (y))).toInt + 1
  private def countingSort (m: Int, c: Array[Int], p: Array[Int], o: Array[Int]): Unit = {
    val cnt:Array[Int] = Array.ofDim (m)
    p.foreach (pi => cnt(c(pi)) = cnt(c(pi)) + 1)
    (1 until m).foreach (i => cnt(i) = cnt(i) + cnt(i-1))
    p.reverseIterator.foreach (pi => {
      cnt(c(pi)) = cnt(c(pi)) - 1
      o(cnt(c(pi))) = pi
    })
  }
  private def build ():Array[Int] = {
    val l = s.length
    var p: Array[Int] = Array.ofDim (l)
    var c = s.map (c => c.toInt).toArray
    var q: Array[Int] = Array.ofDim (l)
    countingSort (sigma, c, (0 until l).toArray, p)
    c(p(0)) = 0
    var m = 0
    for (i <- 1 until l) {
      if (s(p(i)) != s(p(i-1))) {
        m = m + 1
      }
      c(p(i)) = m
    }
    m = m + 1
    var step = 1
    while (step < l) {
      countingSort (m, c, p.map{v => (v + l - step) % l}, q)
      val t1 = p; p = q; q = t1
      q(p(0)) = 0
      m = 0
      for (i <- 1 until l) {
        if (c(p(i)) != c(p(i-1)) || c((p(i) + step) % l) != c((p(i-1) + step) % l)) {
          m = m + 1
        }
        q(p(i)) = m
      }
      m = m + 1
      val t2 = c; c = q; q = t2
      step = 2 * step
    }
    p.slice (1, l)
  }
  val o = build()
  private def lcpTable (a: Array[Int], l: Int, r: Int): Int = {
    if (r - l == 1) {
      a(r)
    } else {
      val m = (l + r) >> 1
      val res = Math.min (lcpTable (a, l, m), lcpTable (a, m, r))
      a(n + 1 + m) = res
      res
    }
  }
  private def lcpBuild () = {
    val q = Array.ofDim[Int](2 * n + 1)
    val r = Array.ofDim[Int](n)
    (0 until n).foreach (i => r(o(i)) = i)
    var l = 0
    for (j <- 0 until n) {
      l = 0.max (l - 1)
      val i = r(j)
      if (i > 0) {
        val k = o(i - 1)
        while ((j + l < n) && (k + l < n) && s(j + l) == s(k + l)) {
          l = l + 1
        }
      } else {
        l = 0
      }
      q(i) = l
    }
    lcpTable (q, -1, n)
    q
  }
  val lcp = lcpBuild()
  def find (x: String): SAFindResult = {
    def l (l: Int, r: Int): Int = if (r - l == 1) lcp (r) else lcp (n + 1 + ((l + r) >> 1))
    def pl(l: Int, o: Int): Int = {
      @tailrec
      def loop(k: Int): Int = {
        if(k >= x.length || k + o >= n || x(k) != s(k + o)) k else loop(k+1)
      }
      loop(l)
    }
    @tailrec
    def f (u: Int, lu: Int, v: Int, lv: Int): SAFindResult = {
      if (u + 1 >= v) {
        SAInterval(u, v)
      } else {
        val m = (u + v) >> 1
        if (lu <= l (m, v) && l (m, v) < lv) {
          f (m, l (m, v), v, lv)
        } else if (lu <= lv && lv < l (m, v)) {
          f (u, lu, m, lv)
        } else if (lv <= l (u, m) && l (u, m) < lu) {
          f (u, lu, m, l (u, m))
        } else if (lv <= lu && lu < l (u, m)) {
          f (m, lu, v, lv)
        } else {
          val om = o(m)
          val lm = n - om
          val l = pl(lu.max(lv), o(m))
          if (l == x.length && l == lm) {
            SASuffix(m)
          } else if ((l == lm) || (l != x.length && s(om+l) < x(l))) {
            f (m, l, v, lv)
          } else {
            f (u, lu, m, l)
          }
        }
      }
    }
    f (-1, 0, n, 0)
  }
}

File: rust/src/suffix_array.rs
pub struct SuffixArray {
  s: Vec<u8>,
  o: Vec<u32>,
  n: usize,
}

fn counting_sort(m: usize, c: &Vec<u32>, p: &Vec<u32>, o: &mut Vec<u32>) {
  let mut cnt = vec![0; m];
  for v in p {
    cnt[c[*v as usize] as usize] += 1;
  }
  for i in 1..m {
    cnt[i] += cnt[i - 1];
  }
  for pi in p.iter().rev() {
    let u = c[*pi as usize] as usize;
    cnt[u] -= 1;
    o[cnt[u]] = *pi;
  }
}

fn lcp_probe(lcp: &Vec<u32>, n: usize, l: isize, r: isize) -> u32 {
  if r - l == 1 {
    lcp[r as usize]
  } else {
    let m = (l + r) >> 1;
    lcp[n + 1 + m as usize]
  }
}

fn lcp_build(lcp: &mut Vec<u32>, n: usize, l: isize, r: isize) -> u32 {
  if r - l == 1 {
    lcp[r as usize]
  } else {
    let m = (l + r) >> 1;
    let v = lcp_build(lcp, n, l, m).min(lcp_build(lcp, n, m, r));
    lcp[n + 1 + m as usize] = v;
    v
  }
}

impl SuffixArray {
  pub fn new(s: &String) -> Self {
    let mut s = s.as_bytes().to_vec();
    s.push(0);
    let n = s.len();
    let mut c = s.iter().map(|q| *q as u32).collect::<Vec<_>>();
    let a = (1 + *c.iter().max().unwrap()) as usize;
    let mut p = vec![0u32; n];
    counting_sort(a, &c, &(0..n as u32).collect::<Vec<_>>(), &mut p);
    c[p[0] as usize] = 0;
    let mut m = 0;
    for i in 1..n {
      let pi = p[i] as usize;
      if s[pi] != s[p[i - 1] as usize] {
        m += 1;
      }
      c[pi] = m;
    }
    m += 1;
    let mut t = vec![0; n];
    let mut step = 1;
    while step < n {
      let s = step as u32;
      for q in &mut p {
        if *q < s {
          *q += n as u32;
        }
        *q -= s;
      }
      counting_sort(m as usize, &c, &p, &mut t);
      std::mem::swap(&mut t, &mut p);
      t[p[0] as usize] = 0;
      m = 0;
      let mut pj = p[0] as usize;
      for i in 1..n {
        let pi = p[i] as usize;
        if c[pi] != c[pj] || c[(pi + step) % n] != c[(pj + step) % n] {
          m += 1;
        }
        t[pi] = m;
        pj = pi;
      }
      m += 1;
      std::mem::swap(&mut c, &mut t);
      step <<= 1;
    }
    debug_assert_eq!(p[0] as usize, n - 1);
    Self {
      s,
      o: p[1..].to_vec(),
      n: n - 1,
    }
  }
  pub fn compute_lcp(&self) -> Vec<u32> {
    let n = self.n;
    let mut lcp = vec![0; 2 * n + 1];
    let mut r = vec![0; n];
    for i in 0..n {
      r[self.o[i] as usize] = i as u32;
    }
    let mut l = 0;
    for j in 0..n {
      if l > 0 {
        l -= 1;
      }
      let i = r[j] as usize;
      if i > 0 {
        let k = self.o[i - 1] as usize;
        while j + l < n && k + l < n && self.s[j + l] == self.s[k + l] {
          l += 1;
        }
      } else {
        l = 0;
      }
      lcp[i] = l as u32;
    }
    debug_assert_eq!(lcp[n], 0);
    lcp_build(&mut lcp, n, -1, n as isize);
    lcp
  }
}
